File: packages/health_connector_hc_android/lib/src/mappers/health_record_mappers/elevation_gained_record_mapper.dart
import 'package:health_connector_core/health_connector_core_internal.dart';
import 'package:health_connector_hc_android/src/mappers/metadata_mappers/metadata_mapper.dart';
import 'package:health_connector_hc_android/src/pigeon/health_connector_hc_android_api.g.dart';
import 'package:meta/meta.dart';

/// Converts [ElevationGainedRecord] to [ElevationGainedRecordDto].
@sinceV3_1_0
@internal
extension ElevationGainedRecordToDto on ElevationGainedRecord {
  ElevationGainedRecordDto toDto() {
    return ElevationGainedRecordDto(
      id: id.value,
      startTime: startTime.millisecondsSinceEpoch,
      endTime: endTime.millisecondsSinceEpoch,
      metadata: metadata.toDto(),
      meters: elevation.inMeters,
      startZoneOffsetSeconds: startZoneOffsetSeconds,
      endZoneOffsetSeconds: endZoneOffsetSeconds,
    );
  }
}

/// Converts [ElevationGainedRecordDto] to [ElevationGainedRecord].
@sinceV3_1_0
@internal
extension ElevationGainedRecordDtoToDomain on ElevationGainedRecordDto {
  ElevationGainedRecord toDomain() {
    return ElevationGainedRecord.internal(
      id: HealthRecordId(id!),
      startTime: DateTime.fromMillisecondsSinceEpoch(startTime, isUtc: true),
      endTime: DateTime.fromMillisecondsSinceEpoch(endTime, isUtc: true),
      metadata: metadata.toDomain(),
      elevation: Length.meters(meters),
      startZoneOffsetSeconds: startZoneOffsetSeconds,
      endZoneOffsetSeconds: endZoneOffsetSeconds,
    );
  }
}

File: packages/health_connector_hc_android/android/src/main/kotlin/com/phamtunglam/health_connector_hc_android/mappers/health_record_mappers/ElevationGainedRecordMapper.kt
package com.phamtunglam.health_connector_hc_android.mappers.health_record_mappers

import androidx.health.connect.client.records.ElevationGainedRecord
import androidx.health.connect.client.units.Length
import com.phamtunglam.health_connector_hc_android.mappers.metadata_mappers.toDto
import com.phamtunglam.health_connector_hc_android.mappers.metadata_mappers.toHealthConnect
import com.phamtunglam.health_connector_hc_android.pigeon.ElevationGainedRecordDto
import java.time.Instant
import java.time.ZoneOffset

/**
 * Converts a Health Connect [ElevationGainedRecord] object to a [ElevationGainedRecordDto].
 */
internal fun ElevationGainedRecord.toDto(): ElevationGainedRecordDto = ElevationGainedRecordDto(
    id = metadata.id,
    startTime = startTime.toEpochMilli(),
    endTime = endTime.toEpochMilli(),
    startZoneOffsetSeconds = startZoneOffset?.totalSeconds?.toLong(),
    endZoneOffsetSeconds = endZoneOffset?.totalSeconds?.toLong(),
    metadata = metadata.toDto(),
    meters = elevation.inMeters,
)

/**
 * Converts a [ElevationGainedRecordDto] to a Health Connect [ElevationGainedRecord] object.
 */
internal fun ElevationGainedRecordDto.toHealthConnect(): ElevationGainedRecord =
    ElevationGainedRecord(
        elevation = Length.meters(meters),
        startTime = Instant.ofEpochMilli(startTime),
        endTime = Instant.ofEpochMilli(endTime),
        startZoneOffset = startZoneOffsetSeconds?.let { ZoneOffset.ofTotalSeconds(it.toInt()) },
        endZoneOffset = endZoneOffsetSeconds?.let { ZoneOffset.ofTotalSeconds(it.toInt()) },
        metadata = metadata.toHealthConnect(id),
    )
